colossalai/kernel/cuda_native/csrc/attention_infer_kernels/linear/linear_op.cpp

```diff
@@ -0,0 +1,206 @@
+#include <vector>
+#include <cuda_fp16.h>
+#include <cublas_v2.h>
+#include <torch/extension.h>
+#include <ATen/cuda/CUDAContext.h>
+
+
+class CublasHandle
+{
+public:
+    static CublasHandle& instance()
+    {
+        static CublasHandle handle;
+        return handle;
+    }
+
+    cublasHandle_t get() const
+    {
+        return handle;
+    }
+
+    CublasHandle(CublasHandle const&) = delete;
+    void operator=(CublasHandle const&)  = delete;
+
+private:
+    cublasHandle_t handle;
+
+    CublasHandle()
+    {
+        cublasStatus_t stat = cublasCreate(&handle);
+        if (stat != CUBLAS_STATUS_SUCCESS)
+        {
+            printf("cuBLAS initialization error: %d\n", stat);
+            exit(stat);
+        }
+    }
+
+    ~CublasHandle()
+    {
+        cublasDestroy(handle);
+    }
+};
+
+class CudaStream {
+public:
+    // Get the singleton instance
+    static CudaStream& instance() {
+        static CudaStream instance;
+        return instance;
+    }
+
+    // Get the cudaStream_t
+    cudaStream_t get() const {
+        return stream;
+    }
+
+private:
+    // The cudaStream_t object
+    cudaStream_t stream;
+
+    // Private constructor and destructor
+    CudaStream() {
+        cudaError_t err = cudaStreamCreate(&stream);
+        if (err != cudaSuccess) {
+            printf("cuda stream initialization error");
+            exit(-1);
+        }
+    }
+
+    ~CudaStream() {
+        cudaStreamDestroy(stream);
+    }
+
+    // Delete copy and assignment constructors
+    CudaStream(const CudaStream&) = delete;
+    CudaStream& operator=(const CudaStream&) = delete;
+};
+
+
+
+
+void dense_layer_fp32_kernel(const float *in, const float *weight, float *out, const int M,
```

Comment:
As we discussed, this kernel is less efficient than pytorch's. Did we optimize this one? If not, we may not add this kernel to colossalai.

Reply:
yes, I think we can delete this since linear implementation is not good.